client/src/main/java/io/atomix/copycat/client/session/ClientSession.java

```diff
@@ -580,12 +589,30 @@ private ClientSession resetConnection() {
    */
   private ClientSession resetMembers() {
     if (connectMembers.isEmpty() || connectMembers.size() < members.size() - 1) {
-      connectMembers = connectionStrategy.getConnections(leader, new ArrayList<>(members));
+      connectMembers = new ArrayList<>(members);
     }
     return this;
   }
 
   /**
+   * Selects the next member to which to send a request.
+   */
+  private Address selectMember() {
+    // Remove the next member that meets the strategy from the members list.
+    Iterator<Address> iterator = connectMembers.iterator();
+    while (iterator.hasNext()) {
+      Address member = iterator.next();
+      if (connectionStrategy.connect(member, leader)) {
+        iterator.remove();
```

Comment:
Sorry you were talking about establishing a session, but I think the above also needs to be figured out. The point about what happens if a client can't initially establish a session we've debated in the past. When initializing the session, it does indeed currently only try each member once. I think failing the initial `open` call if a client can't immediately establish a session is intuitive. But there are definitely cases where it wouldn't be right to fail the `open` because a session couldn't be establish, e.g. when there's just no leader. So, one option is to add a strategy for opening a session. The `ConnectionStrategy`, which may be poorly named, dictates the servers to favor. In other words, for read-heavy workloads, connect to followers, and for write-heavy connect to the leader. I guess a strategy for opening a session would allow retries with exponential back off.

But we could also use the server response errors. If `NO_LEADER_ERROR` is returned by all servers, wait for a leader to be elected.

Maybe I'll just put all those in another strategy.